src/canvas.ts

```diff
@@ -0,0 +1,101 @@
+/**
+ *
+ * Part of the MLP r/place Project, under the Apache License v2.0 or ISC.
+ * SPDX-License-Identifier: Apache-2.0 OR ISC
+ * SPDX-FileCopyrightText: Copyright CONTRIBUTORS.md
+ *
+ **
+ *
+ * @file General purpose canvas handling, abstracts out site specific details.
+ *
+ **/
+
+export function waitForDocumentLoad() {
+  return new Promise<void>((resolve) => {
+    if (document.readyState == 'complete')
+      resolve();
+    const readystatechangeHandler = function() {
+      if (document.readyState == 'complete') {
+        document.removeEventListener('readystatechange', readystatechangeHandler);
+        resolve();
+      }
+    };
+    document.addEventListener('readystatechange', readystatechangeHandler);
+  });
+}
+
+function findCanvasCandidates(root: Element | ShadowRoot, out: HTMLCanvasElement[]) {
+  if (root instanceof HTMLCanvasElement)
+    out.push(root);
+  if (root instanceof HTMLElement && root.shadowRoot)
+    findCanvasCandidates(root.shadowRoot, out);
+  for (const child of root.children)
+    findCanvasCandidates(child, out);
+}
+
+function rankCanvasCandidates(inout: HTMLCanvasElement[]) {
+  inout.sort((a, b) => {
+    // Compare size.
+    const size = (a.width * a.height) - (b.width * b.height);
+    if (size != 0)
+      return -size;
+    
+    // Compare opacity.
+    const opacity = Number(window.getComputedStyle(a).opacity) -
+                    Number(window.getComputedStyle(b).opacity);
```

Comment:
What is the intention behind this?
In case there are multiple canvases overlayed over each other?
Perhaps it's also worth checking if a canvas is empty? Empty canvases are transparent AFAIK.

Reply:
Some sites like pxls.space have a bunch of overlayed canvases. I think the only way to check if it's empty is to actually grab the data and look.